plat/common/w_xor_x.c

```diff
@@ -0,0 +1,122 @@
+ * Copyright (c) 2022, OpenSynergy GmbH. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <uk/config.h>
+#include <uk/plat/memory.h>
+#include <uk/plat/paging.h>
+#include <uk/print.h>
+
+#ifdef CONFIG_ARCH_ARM_64
+#include <arm/arm64/cpu.h>
+#include <arm/arm64/cpu_defs.h>
+#endif /* CONFIG_ARCH_ARM_64 */
+
+#ifdef CONFIG_UKPLAT_MEMRNAME
+#define WXORX_REGION_NAME d.name
+#else /* !CONFIG_UKPLAT_MEMRNAME */
+#define WXORX_REGION_NAME "memory range"
+#endif /* CONFIG_UKPLAT_MEMRNAME */
+
+#ifdef CONFIG_ARCH_ARM_64
+#define enable_wxn() ({				\
+	__u64 reg;				\
+	reg = SYSREG_READ64(SCTLR_EL1);		\
+	reg |= SCTLR_WXN;			\
+	SYSREG_WRITE64(SCTLR_EL1, reg);		\
+	isb();					\
+})
+#endif /* CONFIG_ARCH_ARM_64 */
+
+void __weak enforce_w_xor_x(void)
+{
+	struct ukplat_memregion_desc d;
+	void *base;
+	void *end;
+	unsigned long prot, pages;
+	int rc, i = 0;
+
+	while ((rc = ukplat_memregion_get(i++, &d)) >= 0) {
+ #ifdef CONFIG_ARCH_ARM_64
+		/* Skip RW regions. These will be protected by WXN */
+		if (d.flags & UKPLAT_MEMRF_WRITABLE)
+			continue;
+#endif /* CONFIG_ARCH_ARM64 */
+
+		/* We expect platforms to provide sane sections,
+		 * that is, if multiple sections reside in the same
+		 * page, they should have consistent protections.
+		 */
+		base = (void *)ALIGN_DOWN((unsigned long)d.base, PAGE_SIZE);
+		end  = (void *)ALIGN_UP((unsigned long)d.base + d.len,
+			PAGE_SIZE);
+		pages = DIV_ROUND_UP(end - base, PAGE_SIZE);
```

Comment:
To be on the safe side of the C standard, it might be better to use `__uptr` for `base` and `end`, as subtracting pointers is only defined for elements in the same array 🤓
```suggestion
		base = ALIGN_DOWN((__uptr)d.base, PAGE_SIZE);
		end  = ALIGN_UP((__uptr)d.base + d.len,
			PAGE_SIZE);
		pages = (end - base) / PAGE_SIZE;
```